tests/queries/0_stateless/02500_remove_redundant_distinct.sh

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+OPTIMIZATION_SETTING="query_plan_remove_redundant_distinct"
+DISABLE_OPTIMIZATION="SET allow_experimental_analyzer=1;SET $OPTIMIZATION_SETTING=0;SET optimize_duplicate_order_by_and_distinct=0"
+ENABLE_OPTIMIZATION="SET allow_experimental_analyzer=1;SET $OPTIMIZATION_SETTING=1;SET optimize_duplicate_order_by_and_distinct=0"
+
+echo "-- Disabled $OPTIMIZATION_SETTING"
+query="SELECT DISTINCT *
+FROM
+(
+    SELECT DISTINCT *
+    FROM
+    (
+        SELECT DISTINCT *
+        FROM numbers(3)
+    )
+)"
+
+$CLICKHOUSE_CLIENT -nq "$DISABLE_OPTIMIZATION;EXPLAIN $query"
+
+function run_query {
+    echo "-- query"
+    echo "$1"
+    echo "-- explain"
+    $CLICKHOUSE_CLIENT -nq "$ENABLE_OPTIMIZATION;EXPLAIN $1"
+    echo "-- execute"
+    $CLICKHOUSE_CLIENT -nq "$ENABLE_OPTIMIZATION;$1"
+}
+
+echo "-- Enabled $OPTIMIZATION_SETTING"
+echo "-- DISTINCT is only in most inner subquery"
+run_query "$query"
+
+echo "-- do _not_ remove DISTINCT after UNION"
```

Comment:
Since there is special handling for `UNION`, is special handling also needed for `INTERSECT`?

Reply:
It's [here](https://github.com/ClickHouse/ClickHouse/pull/44176/files#diff-f9d3f69f02b009cb46cf780369690700ff9754315793643c4cb8416a38ab118aR54)